runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java

```diff
@@ -369,6 +362,51 @@ public void outputWithTimestamp(OutputT output, Instant timestamp) {
       checkNotNull(combiner, "Combiner passed to createAggregator cannot be null");
       return aggregatorFactory.createAggregatorForDoFn(fn.getClass(), stepContext, name, combiner);
     }
+
+    @Override
+    public BoundedWindow window() {
+      throw new UnsupportedOperationException("Cannot access window.");
+    }
+
+    @Override
+    public Context context(DoFn<InputT, OutputT> doFn) {
```

Comment:
This is the only substantive "extra" parameter. Blending the `DoFn.Context` and `DoFn.ExtraContextFactory` keeps the code diff smaller and reduces a smidge of boilerplate. It has been used elsewhere, so I went with it.